edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/HttpRequestAuthenticator.cs

```diff
@@ -40,94 +39,73 @@ public async Task<HttpAuthResult> AuthenticateAsync(string actorDeviceId, Option
-                {
-                    return new HttpAuthResult(false, Events.AuthenticationFailed("Invalid authorization header count"));
-                }
+                sharedAccessSignature = SharedAccessSignature.Parse(this.iotHubName, authHeader);
+            }
+            catch (Exception ex)
+            {
+                throw new AuthenticationException(Events.AuthenticationFailed($"Cannot parse SharedAccessSignature because of the following error - {ex.Message}"), ex);
+            }
 
-                string authHeader = authorizationQueryParameters.Count == 1
-                    ? authorizationQueryParameters.First()
-                    : authorizationHeaderValues.First();
+            return authHeader;
+        }
 
-                if (!authHeader.StartsWith("SharedAccessSignature", StringComparison.OrdinalIgnoreCase))
-                {
-                    return new HttpAuthResult(false, Events.AuthenticationFailed("Invalid Authorization header. Only SharedAccessSignature is supported."));
-                }
+        async Task<Option<X509Certificate2>> GetClientCertificate(HttpContext context, string actorDeviceId, Option<string> actorModuleId)
+        {
+            X509Certificate2 clientCertificate = await context.Connection.GetClientCertificateAsync();
 
-                try
-                {
-                    SharedAccessSignature sharedAccessSignature = SharedAccessSignature.Parse(this.iotHubName, authHeader);
-                    if (sharedAccessSignature.IsExpired())
-                    {
-                        return new HttpAuthResult(false, Events.AuthenticationFailed("SharedAccessSignature is expired"));
-                    }
-                }
-                catch (Exception ex)
+            string actorId = actorModuleId.Match(mid => $"{actorDeviceId}/{mid}", () => actorDeviceId);
+            try
+            {
+                if (clientCertificate == null)
                 {
-                    return new HttpAuthResult(false, Events.AuthenticationFailed($"Cannot parse SharedAccessSignature because of the following error - {ex.Message}"));
+                    return await this.httpProxiedCertificateExtractor.GetClientCertificate(context);
                 }
-
-                clientCredentials = this.identityFactory.GetWithSasToken(actorDeviceId, actorModuleId.OrDefault(), string.Empty, authHeader, false, Option.None<string>(), authChain);
             }
-
-            IIdentity identity = clientCredentials.Identity;
-
-            if (!await this.authenticator.AuthenticateAsync(clientCredentials))
+            catch (Exception e)
             {
-                return new HttpAuthResult(false, Events.AuthenticationFailed($"Unable to authenticate device with Id {identity.Id}"));
+                throw new AuthenticationException($"Unable to authenticate device with Id {actorId} - {e.Message}", e);
```

Comment:
Is this desirable? If await this.httpProxiedCertificateExtractor.GetClientCertificate(context) throws, then we will fail the request for Auth. What if it is a SAS flow?

Reply:
httpProxiedCertificateExtractor.GetClientCertificate(context) can throw if the header with the certificate was set so I think it should not go to SAS flow in that case.

Reply:
Okay, but in that case we should make sure this ever throws in other cases (the concern is that this throws for non-proxied/sas auth cases).

Reply:
SAS flow should work as before if the certificate is not set in header, which should not be set because proxy is the one setting it, unit tests cover the case. I will test changes without proxy, we tested yesterday with proxy and SAS auth and it worked.